src/util/importJSON.ts

```diff
@@ -0,0 +1,160 @@
+const getRankIncrement = (thoughtsJSON: Block[], state: State, context: Context, destThought: Child, rankStart: number) => {
+  const numContexts = getContextsNum(thoughtsJSON)
+  const destValue = destThought.value
+  const destRank = destThought.rank
+  const next = nextSibling(state, destValue, context, destRank) // paste after last child of current thought
+  const rankIncrement = next ? (next.rank - rankStart) / (numContexts || 1) : 1 // prevent divide by zero
+  return rankIncrement
+}
+
+/** Return start context for saving thoughts. */
+const getStartContext = (thoughtsRanked: Path) => {
+  const importCursor = equalPath(thoughtsRanked, [{ value: EM_TOKEN, rank: 0 }])
+    ? thoughtsRanked
+    : contextOf(thoughtsRanked)
+  return pathToContext(importCursor)
+}
+
+/** Convert JSON to thoughts update. */
+export const importJSON = (state: State, thoughtsRanked: Path, thoughtsJSON: Block[], { skipRoot }: ImportHtmlOptions = { skipRoot: false }) => {
+  const thoughtIndexUpdates: GenericObject<Lexeme> = {}
+  const contextIndexUpdates: GenericObject<ParentEntry> = {}
+  const context = pathToContext(contextOf(thoughtsRanked))
+  const destThought = head(thoughtsRanked)
+  const destEmpty = destThought.value === '' && getThoughts(state, pathToContext(thoughtsRanked)).length === 0
+  const thoughtIndex = { ...state.thoughts.thoughtIndex }
+  const rankStart = getRankAfter(state, thoughtsRanked)
+  const rankIncrement = getRankIncrement(thoughtsJSON, state, context, destThought, rankStart)
+
+  // if the thought where we are pasting is empty, replace it instead of adding to it
+  if (destEmpty) {
+    const thought = getThought(state, '')
+    if (thought && thought.contexts && thought.contexts.length > 1) {
+      thoughtIndexUpdates[hashThought('')] = removeContext(thought, context, headRank(thoughtsRanked))
+      const rootedContext = pathToContext(rootedContextOf(thoughtsRanked))
+      const contextEncoded = hashContext(rootedContext)
+      contextIndexUpdates[contextEncoded] = {
+        ...contextIndexUpdates[contextEncoded],
+        children: getThoughts(state, rootedContext)
+          .filter(child => !equalThoughtRanked(child, destThought)),
+        lastUpdated: timestamp(),
+      }
+    }
+  }
+
+  const lastThoughtFirstLevel = calculateLastThoughtFirstLevel(rankStart, rankIncrement, thoughtsJSON)
+
+  /** Insert the given value at the context. Modifies contextIndex and thoughtIndex. */
+  const insertThought = (value: string, context: Context, rank: number) => {
+    value = value.trim()
+    const id = createId()
+    const rootContext = context.length > 0 ? context : [ROOT_TOKEN]
+    const thoughtNew = addThought(
+      {
+        thoughts: {
+          thoughtIndex
+        }
+      },
+      value,
+      rank,
+      id,
+      rootContext
+    )
+
+    thoughtIndex[hashThought(value)] = thoughtNew
+    thoughtIndexUpdates[hashThought(value)] = thoughtNew
```

Comment:
Let's call `hashThought(value)` once and reuse it (this was in the original code).